typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1316 @@
+        Maybe_shared.set_barrier l0 uniq;
+        m1
+    | Shared _, Borrowed _ -> m0
+    | Maybe_unique l, Borrowed occ ->
+        Maybe_unique.force_shared_multiuse l ~there:occ ~first_or_second:First;
+        Shared occ
+    | Shared _, Maybe_shared _ -> m0
+    | Maybe_unique l0, Maybe_shared l1 ->
+        (* Four cases:
+            Shared;Borrowed = Shared
+            Shared;Shared = Shared
+            Unique;Borrowed = Error
+            Unique;Shared = Error
+
+            As you can see, we need to force the m0 to Shared, and m1 needn't
+            be constrained. The result is always Shard.
+        *)
+        let occ = Maybe_shared.extract_occurrence l1 in
+        Maybe_unique.force_shared_multiuse l0 ~there:occ ~first_or_second:First;
+        Shared occ
+    | Shared _, Shared _ -> m0
+    | Maybe_unique l, Shared occ ->
+        Maybe_unique.force_shared_multiuse l ~there:occ ~first_or_second:First;
+        Shared occ
+    | Shared occ, Maybe_unique l ->
+        Maybe_unique.force_shared_multiuse l ~there:occ ~first_or_second:Second;
+        Shared occ
+    | Maybe_unique l0, Maybe_unique l1 ->
+        let occ1 = Maybe_unique.extract_occurrence l1 in
+        let occ0 = Maybe_unique.extract_occurrence l0 in
+        Maybe_unique.force_shared_multiuse l0 ~there:occ1 ~first_or_second:First;
+        Maybe_unique.force_shared_multiuse l1 ~there:occ0
+          ~first_or_second:Second;
+        Shared (Maybe_unique.extract_occurrence l0)
+end
+
+(** lifting module Usage to trees *)
+module Usage_tree = struct
+
+
+  type t = { children : t Projection.Map.t; usage : Usage.t }
+  (** Represents a tree of usage. Each node records the par on all possible
+     execution paths. As a result, trees such as `S -> U` is valid, even though
+     it would be invalid if it was the result of a single path: using a parent
+     shared and a child uniquely is obviously bad. Howerver, it might be the
+     result of "par"ing multiple path: `S` par `N -> U`, which is valid.
+
+     INVARIANT: children >= parent. For example, having a shared child under a
+     unique parent is nonsense. The invariant is preserved because Usage.par and
+     Usage.seq above is monotone, and Usage_tree.par and Usage_tree.seq here is
+     node-wise. *)
+
+  let rec print_children ppf children =
+    Projection.Map.iter
+      (fun proj child ->
+        Format.fprintf ppf "%a = %a," Projection.print proj print child)
+      children
+
+  and print ppf t =
+    Format.fprintf ppf "%a {%a}" Usage.print t.usage print_children t.children
+
+  module Path = struct
+    type t = Projection.t list
+
+    let child (p : t) (a : Projection.t) : t = p @ [ a ]
+    let root : t = []
+    let _print ppf = Format.pp_print_list Projection.print ppf
+  end
+
+  let leaf usage = { usage; children = Projection.Map.empty }
+  let _unused = leaf Usage.Unused
+
+  (** f must be monotone *)
+  let map ?(on_exception = fun _exn _projs -> assert false) ?(projs=[]) f t =
```

Comment:
The `projs` optional parameter seems to only be used in `map2`. I think it would be better to make it non-optional, call this function `map_with_projs`, define `map` in terms of it, and then only expose `map` in this module's signature.